File: Multimi/Subiect16/subiect16_micronaut_java/src/main/java/com/sd/laborator/VerificareService.java
package com.sd.laborator;

import jakarta.inject.Singleton;
import java.util.ArrayList;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Map;
import java.util.Set;

@Singleton
public class VerificareService {

    public boolean verificaPereche(int a, int b) {
        return a * b == a + b * 3;
    }

    public List<Map<String, Integer>> gasestePerechile(List<Integer> adtA, List<Integer> adtB) {
        List<Map<String, Integer>> rezultate = new ArrayList<>();
        Set<String> vazute = new LinkedHashSet<>();

        for (int a : adtA) {
            for (int b : adtB) {
                String cheie = a + "," + b;
                if (verificaPereche(a, b) && vazute.add(cheie)) {
                    rezultate.add(Map.of("a", a, "b", b));
                }
            }
        }

        return rezultate;
    }
}

File: Multimi/Subiect16/perechi-gradle/src/main/kotlin/com/sd/laborator/VerificareService.kt
package com.sd.laborator

import jakarta.inject.Singleton

/**
 * Serviciu care implementeaza verificarea matematica din enunt.
 *
 * Conditia: a * b == a + b * 3
 *
 * Rezolvare analitica (pentru context):
 *   a*b = a + 3*b
 *   a*b - a = 3*b
 *   a*(b-1) = 3*b
 *   a = 3*b / (b-1)   [cu b != 1]
 *
 * Deci pentru un b dat, a este unic determinat (daca este intreg).
 * Aceasta functie verifica perechea (a, b) din ADT-urile primite.
 */
@Singleton
class VerificareService {

    /**
     * Verifica daca perechea (a, b) satisface a * b == a + b * 3.
     * Aceasta este "functia" mentionata in enunt.
     *
     * @param a element din ADT-ul A
     * @param b element din ADT-ul B
     * @return true daca a * b == a + b * 3
     */
    fun verificaPereche(a: Int, b: Int): Boolean {
        return a * b == a + b * 3
    }

    /**
     * Cauta toate perechile (a, b) din ADT-urile date care satisfac conditia.
     * Perechile duplicate (acelasi (a,b)) sunt incluse o singura data.
     *
     * @param adtA ADT-ul A
     * @param adtB ADT-ul B
     * @return lista de map-uri {"a": ..., "b": ...} cu perechile gasite
     */
    fun gasestePerechile(
        adtA: List<Int>,
        adtB: List<Int>
    ): List<Map<String, Int>> {
        val rezultate = mutableListOf<Map<String, Int>>()
        val vazute = mutableSetOf<Pair<Int, Int>>()

        for (a in adtA) {
            for (b in adtB) {
                val pereche = Pair(a, b)
                if (verificaPereche(a, b) && pereche !in vazute) {
                    rezultate.add(mapOf("a" to a, "b" to b))
                    vazute.add(pereche)
                }
            }
        }

        return rezultate
    }
}
